testsuite/tests/typing-jkind-bounds/row.ml

```diff
@@ -0,0 +1,278 @@
+[%%expect{|
+type 'a v = [ `C of 'a | `D of 'a -> 'a | `E of 'a option ]
+|}]
+
+type 'a w : value mod contended with 'a = [ 'a t | 'a v ]
+[%%expect{|
+type 'a w = [ `A of 'a | `C of 'a | `D of 'a -> 'a | `E of 'a option ]
+|}]
+
+let cross_contention (x : int t @ contended) = use_uncontended x
+let cross_portability (x : int t @ nonportable) = use_portable x
+let cross_linearity (x : int t @ once) = use_many x
+[%%expect{|
+val cross_contention : int t @ contended -> unit = <fun>
+val cross_portability : int t -> unit = <fun>
+val cross_linearity : int t @ once -> unit = <fun>
+|}]
+
+let don't_cross_unique (x : int t @ aliased) = use_unique x
+[%%expect{|
+Line 1, characters 58-59:
+1 | let don't_cross_unique (x : int t @ aliased) = use_unique x
+                                                              ^
+Error: This value is "aliased" but expected to be "unique".
+|}]
+
+let don't_cross_locality (x : int t @ local) = use_global x
+[%%expect{|
+Line 1, characters 58-59:
+1 | let don't_cross_locality (x : int t @ local) = use_global x
+                                                              ^
+Error: This value escapes its region.
+  Hint: This argument cannot be local,
+  because it is an argument in a tail call.
+|}]
+
+
+let cross_contention (x : int w @ contended) = use_uncontended x
+[%%expect{|
+val cross_contention : int w @ contended -> unit = <fun>
+|}]
+
+let don't_cross_portability (x : int w @ nonportable) = use_portable x
+[%%expect{|
+Line 1, characters 69-70:
+1 | let don't_cross_portability (x : int w @ nonportable) = use_portable x
+                                                                         ^
+Error: This value is "nonportable" but expected to be "portable".
+|}]
+
+(* Quality *)
+
+type 'a record : immutable_data with [ `A of 'a ] = { inner : 'a }
+[%%expect{|
+type 'a record = { inner : 'a; }
+|}]
+
+module type S = sig
+  type 'a polyvar = [ `A of 'a ]
+  type 'a record : immutable_data with 'a polyvar = { inner : 'a }
+end
+[%%expect{|
+module type S =
+  sig type 'a polyvar = [ `A of 'a ] type 'a record = { inner : 'a; } end
+|}]
+
+(* This is why we're able to give closed polymorphic variants best kinds *)
+module type S2 = S with type 'a polyvar = [ `B of int ref ]
+[%%expect{|
+Line 1, characters 17-59:
+1 | module type S2 = S with type 'a polyvar = [ `B of int ref ]
+                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: In this "with" constraint, the new definition of "polyvar"
+       does not match its original definition in the constrained signature:
+       Type declarations do not match:
+         type 'a polyvar = [ `B of int ref ]
+       is not included in
+         type 'a polyvar = [ `A of 'a ]
+       The type "[ `B of int ref ]" is not equal to the type "[ `A of 'a ]"
+       The second variant type does not allow tag(s) "`B"
+|}]
+
+(* When we give open polymorphic variants more precise kinds, we should make sure to give them not-best quality *)
```

Comment:
Very true! I had not thought about this. I do not see a test for bestness of this example, though. (I actually hope it would be sound without effort, though. Probably will end up with an error message mentioning `#row` somewhere, which is regrettable, but good enough for a first release.)

Reply:
Added CRs that hopefully get the message across.